autoPyTorch/evaluation/train_evaluator.py

```diff
@@ -0,0 +1,340 @@
+                if len(Y_test_preds.shape) == 3:
+                    Y_test_preds = np.nanmean(Y_test_preds, axis=0)
+            else:
+                Y_test_preds = None
+
+            self.Y_optimization = Y_targets
+            self.Y_actual_train = Y_train_targets
+
+            self.model = self._get_model()
+
+            status = StatusType.SUCCESS
+            self.logger.debug("In train evaluator fit_predict_and_loss, loss:{}".format(opt_loss))
+            self.finish_up(
+                loss=opt_loss,
+                train_loss=train_loss,
+                opt_pred=Y_optimization_preds,
+                valid_pred=Y_valid_preds,
+                test_pred=Y_test_preds,
+                additional_run_info=additional_run_info,
+                file_output=self.num_folds == 1,
+                status=status,
+            )
+
+    def _fit_and_predict(self, model: BaseEstimator, fold: int, train_indices: Union[np.ndarray, List],
+                         test_indices: Union[np.ndarray, List],
+                         add_model_to_self: bool
+                         ) -> Tuple[np.ndarray, np.ndarray, Optional[np.ndarray], Optional[np.ndarray]]:
+
+        self.indices[fold] = ((train_indices, test_indices))
+
+        # Do only output the files in the case of iterative holdout,
+        # In case of iterative partial cv, no file output is needed
+        # because ensembles cannot be built
+        X = {'train_indices': train_indices,
+             'val_indices': test_indices,
+             'split_id': fold,
+             **self.fit_dictionary}  # fit dictionary
+        y = None
+        fit_and_suppress_warnings(self.logger, model, X, y)
+        self.logger.info("Model fitted, now predicting")
+        (
+            Y_train_pred,
+            Y_opt_pred,
+            Y_valid_pred,
+            Y_test_pred
+        ) = self._predict(
+            model,
+            train_indices=train_indices,
+            test_indices=test_indices,
+        )
+
+        if add_model_to_self:
+            self.model = model
+        else:
+            self.models[fold] = model
+
+        return Y_train_pred, Y_opt_pred, Y_valid_pred, Y_test_pred
+
+    def _predict(self, model: BaseEstimator,
+                 test_indices: Union[np.ndarray, List],
+                 train_indices: Union[np.ndarray, List]
+                 ) -> Tuple[np.ndarray, np.ndarray, Optional[np.ndarray], Optional[np.ndarray]]:
+
+        X_train = self.X_train
+        if isinstance(X_train, pd.DataFrame):
+            X_train = X_train.to_numpy()
+        train_pred = self.predict_function(X_train[train_indices], model,
+                                           self.y_train[train_indices])
+
+        opt_pred = self.predict_function(X_train[test_indices], model,
+                                         self.y_train[train_indices])
+
+        if self.X_valid is not None:
+            valid_pred = self.predict_function(self.X_valid, model,
```

Comment:
What is the difference between `valid_pred` and `opt_pred`?

Reply:
self.X_valid is coming from the datamanager, so in case the user instantiates the dataset by giving a validation set, and a test set then we'll have self.X_valid and self.X_test not None. X_train[test_indices] comes from creating splits in the dataset class, so basically from our resampling strategy.